macros/edr/dbt_artifacts/upload_artifacts_to_table.sql

```diff
@@ -1,18 +1,38 @@
-{% macro upload_artifacts_to_table(table_relation, artifacts, flatten_artifact_callback, truncate_if_on_run_end, should_commit=False) %}
+{% macro upload_artifacts_to_table(table_relation, artifacts, flatten_artifact_callback, append=False, should_commit=False) %}
     {% set flatten_artifact_dicts = [] %}
     {% for artifact in artifacts %}
         {% set flatten_artifact_dict = flatten_artifact_callback(artifact) %}
-        {% if flatten_artifact_dict is not none %}
+        {% if flatten_artifact_dict is not none and elementary.dict_has_nonempty_values(flatten_artifact_dict) %}
             {% do flatten_artifact_dicts.append(flatten_artifact_dict) %}
         {% endif %}
     {% endfor %}
-    {% if truncate_if_on_run_end and elementary.is_on_run_end() %}
-      {% do dbt.truncate_relation(table_relation) %}
+
+    {% if append %}
+        {# In append mode, just insert, and no need to be atomic #}
+        {% do elementary.insert_rows(table_relation, flatten_artifact_dicts, should_commit, elementary.get_config_var('dbt_artifacts_chunk_size')) %}
+    {% else %}
+        {% do elementary.overwrite_artifacts_table(table_relation, flatten_artifact_dicts, should_commit) %}
     {% endif %}
+{% endmacro %}
+
+{% macro overwrite_artifacts_table(table_relation, flatten_artifact_dicts, should_commit) %}
+    {% do return(adapter.dispatch('overwrite_artifacts_table', 'elementary')(table_relation, flatten_artifact_dicts, should_commit)) %}
+{% endmacro %}
+
+{% macro default__overwrite_artifacts_table(table_relation, flatten_artifact_dicts, should_commit) %}
+    {# First upload everything to a temp table #}
+    {% set temp_table_relation = dbt.make_temp_relation(table_relation) %}
+    {% do elementary.create_table_like(temp_table_relation, table_relation, temporary=True) %}
+    {% do elementary.insert_rows(temp_table_relation, flatten_artifact_dicts, should_commit, elementary.get_config_var('dbt_artifacts_chunk_size')) %}
+
+    {# Now atomically replace the data #}
+    {% do elementary.replace_data_with_table_contents(table_relation, temp_table_relation) %}
```

Comment:
Why replace the data and not rename the relation like the materialization does?
It's much more cheap.

Reply:
@elongl so I read somewhere that recreating the table might cause views defined over the table to stop working. But actually I didn't verify that.
WDYT?

Reply:
It may be better though to rename like the materialization does, just need to make sure I can rename under a transaction in all platforms (as it requires two renames)

Reply:
Given that's what dbt do for materlializing models, I feel comfortable doing the same.
I think that `replace_data` would be more expensive in terms of performance and storage (data duplication) on larger projects.

Reply:
So what I'm worried about is that unlike dbt, I'm not doing this as a part of a materialization
It can happen in any dbt run
So not sure I want to risk that views depending on the artifacts will stop working...
I can check though that this is indeed what happens.

Reply:
(I mean that with regular models it's usually clear that if you change a model you have to run models that depend on it too)